### PR TITLE
Delete unused branch plugin: Makes GetBranchNames() support worktrees

### LINKED ISSUE #11375: Use of git worktrees breaks "Delete obsolete branches" plugin
### Environment

- Git Extensions 4.2.0.17591
- Build bf814d4f1c6f178eacef38d592b700f9f2519dcb
- Git 2.43.0.windows.1
- Microsoft Windows NT 10.0.19045.0
- .NET 6.0.25
- DPI 96dpi (no scaling)
- Portable: False

### Issue description

Exit code: 128
Command: C:\Program Files\Git\bin\git.exe
Arguments: -c log.showSignature=false log --pretty=%ci
%an
%s + main^1..+ main
Working directory: D:\_projects\XXXX\main\

```
GitExtUtils.ExternalOperationException: fatal: ambiguous argument '+': unknown revision or path not in the working tree.
Use '--' to separate paths from revisions, like this:
'git <command> [<revision>...] -- [<file>...]'
 ---> System.Exception: fatal: ambiguous argument '+': unknown revision or path not in the working tree.
Use '--' to separate paths from revisions, like this:
'git <command> [<revision>...] -- [<file>...]'
   --- End of inner exception stack trace ---
   at async Task<string> GitCommands.ExecutableExtensions.GetOutputAsync(IExecutable executable, ArgumentString arguments, byte[] input, Encoding outputEncoding, CommandCache cache, bool stripAnsiEscapeCodes) in C:/projects/gitextensions/GitCommands/Git/ExecutableExtensions.cs:line 129
   at void Microsoft.VisualStudio.Threading.JoinableTask.CompleteOnCurrentThread()
   at IEnumerable<Branch> GitExtensions.Plugins.DeleteUnusedBranches.DeleteUnusedBranchesForm.GetObsoleteBranches(RefreshContext context, string curBranch)+MoveNext() in C:/projects/gitextensions/Plugins/DeleteUnusedBranches/DeleteUnusedBranchesForm.cs:line 103
   at void System.Collections.Generic.List<T>.InsertRange(int index, IEnumerable<T> collection)
   at void GitUI.SortableBindingList<T>.AddRange(IEnumerable<T> items) in C:/projects/gitextensions/GitUI/SortableBindingList.cs:line 20
   at async Task GitExtensions.Plugins.DeleteUnusedBranches.DeleteUnusedBranchesForm.RefreshObsoleteBranchesAsync() in C:/projects/gitextensions/Plugins/DeleteUnusedBranches/DeleteUnusedBranchesForm.cs:line 347
   at async void GitUI.TaskManager.InvokeAndForget(Control control, Func<Task> asyncAction, CancellationToken cancellationToken)+(?) => { } [1] in C:/projects/gitextensions/GitExtUtils/GitUI/TaskManager.cs:line 110
   at async Task GitUI.TaskManager.HandleExceptionsAsync(Func<Task> asyncAction, Func<Exception, Task> handleExceptionAsync) in C:/projects/gitextensions/GitExtUtils/GitUI/TaskManager.cs:line 38
```

Opening the "Delete obsolete branches" dialog window directly results in this exception.

### Steps to reproduce

Select the "Plugins -> Delete obsolete branches" menu item.

### Did this work in previous version of GitExtensions?

Yes, but only partly. The search for obsolete branches took forever (see [my comment](https://github.com/gitextensions/gitextensions/issues/10662#issuecomment-1784741453) on the corresponding GitHub issue).

### Diagnostics

When using a worktree, branch checked out in a worktree is prefixed by a "+" in the output of the `git branch --list --merged HEAD` command and as a consequence is badly parsed and a wrong name is used in the following `git log` command run.

### PR DESCRIPTION
**Pull request by pmiossec**
by parsing correctly the output of the command
`git branch --list --merged HEAD`
that contains "+ " prefix in front of branch names checked out in git worktrees

+ add support of some other edge cases

Fixes #11375

+ makes git log command a little more resilient by appending '--' like done in #11277 (if it will fail in the future, it will be with a little better error message)

### Before

image

### After

image

## Test methodology 

- unit tests
- manual

## Test environment(s) 

- GIT 
- Windows 

## Merge strategy

I agree that the maintainer squash merge this PR (if the commit message is clear).

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

### CODE REVIEW

**Comment by pmiossec** [comment]
@jonnydee You could test the fix in the portable version built and please confirm that it solves the issue:

https://ci.appveyor.com/project/gitextensions/gitextensions/builds/48578287/artifacts

**Review by mstv — APPROVED** [review]
looks sensible, have only run the changed "git log" command

File: Plugins/DeleteUnusedBranches/GitBranchOutputCommandParser.cs [hunk 1]
```diff
@@ -9,11 +9,12 @@ public IEnumerable<string> GetBranchNames(string commandOutput)
                 yield break;
             }
 
-            foreach (string line in commandOutput.Split('\n').Where(line => !string.IsNullOrWhiteSpace(line)))
+            foreach (string line in commandOutput.Split('\n').Where(line => !string.IsNullOrWhiteSpace(line) && line.Length > 2))
             {
-                string branchName = line.Trim('*', ' ', '\n', '\r');
+                // Removing first 2 chars: 1st is '*' for current branch or '+' for a worktree branch followed by a space
+                string branchName = line.Substring(2).Trim(' ', '\n', '\r');
```
Comment (gerhardol) [inline Plugins/DeleteUnusedBranches/GitBranchOutputCommandParser.cs:15]: range operator, most calls are changed to this
```suggestion
                string branchName = line[2..].Trim(' ', '\n', '\r');
```

File: Plugins/DeleteUnusedBranches/GitBranchOutputCommandParser.cs [hunk 2]
```diff
@@ -9,11 +9,12 @@ public IEnumerable<string> GetBranchNames(string commandOutput)
...
 
-                if (branchName != "HEAD")
+                if (branchName != "HEAD" && !branchName.StartsWith("(HEAD") && branchName.Length != 0)
```
Comment (gerhardol) [inline Plugins/DeleteUnusedBranches/GitBranchOutputCommandParser.cs:17]: ?
```suggestion
                if (branchName != "HEAD" && !branchName.StartsWith("(") && branchName.Length != 0)
```

File: UnitTests/Plugins/DeleteUnusedBranches.Tests/GitBranchOutputCommandParserTests.cs [hunk 3]
```diff
@@ -28,12 +28,28 @@ public void GetBranchNames_should_return_parse_correctly()
             string commandOutput = @"  feature/branch-1
 * fix/branch-2
   master
+  HEAD
+* (HEAD detached at dce3f9f)
+  +_strange_branch_name_but_valid
```
Comment (gerhardol) [inline UnitTests/Plugins/DeleteUnusedBranches.Tests/GitBranchOutputCommandParserTests.cs:33]: ?
```suggestion
+ _strange_branch_name_but_valid
```

File: UnitTests/Plugins/DeleteUnusedBranches.Tests/GitBranchOutputCommandParserTests.cs [hunk 4]
```diff
@@ -28,12 +28,28 @@ public void GetBranchNames_should_return_parse_correctly()
...
                    .Should()
-                   .BeEquivalentTo("feature/branch-1", "fix/branch-2", "master");
+                   .BeEquivalentTo("feature/branch-1", "fix/branch-2", "master", "+_strange_branch_name_but_valid");
+        }
+
+        [TestCase]
+        public void GetBranchNames_should_support_worktrees()
+        {
+            string commandOutput = @"+ branch_in_worktree
+*  master
```
Comment (gerhardol) [inline UnitTests/Plugins/DeleteUnusedBranches.Tests/GitBranchOutputCommandParserTests.cs:46]: Should it be 2 spaces?
```suggestion
* master
```

**Review by gerhardol — APPROVED** [review]

**Review by RussKie — COMMENTED** [review]
👍 modulo the outstanding feedback

**Comment by RussKie** [comment]
Should we patch 4.2?

**Comment by jonnydee** [comment]
> @jonnydee You could test the fix in the portable version built and please confirm that it solves the issue:
> 
> https://ci.appveyor.com/project/gitextensions/gitextensions/builds/48578287/artifacts

@pmiossec I can confirm that the fix solves the issue. That was pretty fast, thank you very much!

**Comment by jonnydee** [comment]
> Should we patch 4.2?

Yes, please!

**Comment by pmiossec** [comment]
Fixed everything (I think). Rebased on and targeted `release/4.0`.

**Comment by RussKie** [comment]
Thank you

**Pull request merged**